### PR TITLE
[Summer of code] Stand alone a new controller module

### PR DESCRIPTION
**Pull request by hzh0425**
## What is the purpose of the change

tracking issue: https://github.com/apache/rocketmq/issues/4330

## Brief changelog

Add a new module -- controller.
Make the controller independent from name-srv and can be deployed independently

## Verifying this change

Follow this checklist to help us incorporate your contribution quickly and easily. Notice, `it would be helpful if you could finish the following 5 checklist(the last one is not necessary)before request the community to review your PR`.

- [x] Make sure there is a [Github issue](https://github.com/apache/rocketmq/issues) filed for the change (usually before you start working on it). Trivial changes like typos do not require a Github issue. Your pull request should address just this issue, without pulling in other changes - one PR resolves one issue. 
- [x] Format the pull request title like `[ISSUE #123] Fix UnknownException when host config not exist`. Each commit in the pull request should have a meaningful subject line and body.
- [x] Write a pull request description that is detailed enough to understand what the pull request does, how, and why.
- [x] Write necessary unit-test(over 80% coverage) to verify your logic correction, more mock a little better when cross module dependency exist. If the new feature or significant change is committed, please remember to add integration-test in [test module](https://github.com/apache/rocketmq/tree/master/test).
- [x] Run `mvn -B clean apache-rat:check findbugs:findbugs checkstyle:checkstyle` to make sure basic checks pass. Run `mvn clean install -DskipITs` to make sure unit-test pass. Run `mvn clean test-compile failsafe:integration-test`  to make sure integration-test pass.
- [ ] If this contribution is large, please file an [Apache Individual Contributor License Agreement](http://www.apache.org/licenses/#clas).

### CODE REVIEW

File: controller/src/main/java/org/apache/rocketmq/controller/impl/DefaultBrokerHeartbeatManager.java [hunk 1]
```diff
@@ -0,0 +1,225 @@
...
+import org.apache.rocketmq.common.constant.LoggerName;
+import org.apache.rocketmq.common.namesrv.ControllerConfig;
+import org.apache.rocketmq.controller.BrokerHeartbeatManager;
+import org.apache.rocketmq.logging.InternalLogger;
+import org.apache.rocketmq.logging.InternalLoggerFactory;
+import org.apache.rocketmq.remoting.common.RemotingUtil;
+
+public class DefaultBrokerHeartbeatManager implements BrokerHeartbeatManager {
+    private static final InternalLogger log = InternalLoggerFactory.getLogger(LoggerName.CONTROLLER_LOGGER_NAME);
+    private static final long DEFAULT_BROKER_CHANNEL_EXPIRED_TIME = 1000 * 60 * 2;
```
Comment (RongtongJin) [inline controller/src/main/java/org/apache/rocketmq/controller/impl/DefaultBrokerHeartbeatManager.java:39]: The default time is too long. In nameserver, the registration interval is 30s by default, so the timeout time is 120s

File: controller/src/main/java/org/apache/rocketmq/controller/impl/DefaultBrokerHeartbeatManager.java [hunk 2]
```diff
@@ -0,0 +1,225 @@
...
+                ", heartbeatTimeoutMillis=" + heartbeatTimeoutMillis +
+                ", channel=" + channel +
+                '}';
+        }
+    }
+
+    /**
+     * broker address information
+     */
+    static class BrokerAddrInfo {
```
Comment (RongtongJin) [inline controller/src/main/java/org/apache/rocketmq/controller/impl/DefaultBrokerHeartbeatManager.java:177]: 可以把这个类提取到common模块，controller和namesrv一起复用

File: controller/src/main/java/org/apache/rocketmq/controller/impl/DefaultBrokerHeartbeatManager.java [hunk 3]
```diff
@@ -0,0 +1,225 @@
...
+            long last = info.lastUpdateTimestamp;
+            long timeoutMillis = info.heartbeatTimeoutMillis;
+            return (last + timeoutMillis) >= System.currentTimeMillis();
+        }
+        return false;
+    }
+
+    static class BrokerLiveInfo {
+        private final String brokerName;
+        private final long brokerId;
```
Comment (RongtongJin) [inline controller/src/main/java/org/apache/rocketmq/controller/impl/DefaultBrokerHeartbeatManager.java:149]: 这里为什么需要brokerId？

Comment (hzh0425) [reply controller/src/main/java/org/apache/rocketmq/controller/impl/DefaultBrokerHeartbeatManager.java:149]: brokerInactive 的时候需要判断该 broker 是否为 Master, 因为 controller 单独部署, 所以需要保存 brokerId.

**Comment by hzh0425** [comment]
@RongtongJin maybe this pr can be merged

**Review by RongtongJin — APPROVED** [review]

**Pull request merged**